octobot_trading/modes/channel/abstract_mode_producer.py

```diff
@@ -264,21 +278,36 @@ async def finalize(self, exchange_name: str,
             # Do nothing if not its exchange
             return
 
-        with self.trading_mode_trigger():
-            async with self.trading_mode.remote_signal_publisher(symbol):
-                await self.set_final_eval(matrix_id=matrix_id,
-                                          cryptocurrency=cryptocurrency,
-                                          symbol=symbol,
-                                          time_frame=time_frame)
+        async with self.trading_mode_trigger(), self.trading_mode.remote_signal_publisher(symbol):
+            await self.set_final_eval(matrix_id=matrix_id,
+                                      cryptocurrency=cryptocurrency,
+                                      symbol=symbol,
+                                      time_frame=time_frame)
 
-    @contextlib.contextmanager
-    def trading_mode_trigger(self):
+    @contextlib.asynccontextmanager
+    async def trading_mode_trigger(self):
         try:
+            try:
+                if self.exchange_manager.is_backtesting and self._is_first_call:
+                    # reset_exchange_init_data done at first trading mode call in backtesting
+                    await self.trading_mode.reset_exchange_init_data()
+                    self._is_first_call = False
```

Comment:
I don't understand why we have to do that

Reply:
In backtesting, we can't wait for init events in the same way we do in live. In live when init events are triggered, we call `reset_exchange_init_data` which saves candles and other things. We need a way to know when to call this `reset_exchange_init_data` in backtesting and we can't really rely on these event (or we need to use an independant async task that would not be managed by the backtesting engine). The cleanest way I found was this call but it's not the perfect solution. Once we have a proper storage system we will not to do this anymore as it will be the storage system that will call what is currently called by `reset_exchange_init_data` and it will do it at the right moment.
I'm adding a comment to remember to change this